sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java

```diff
@@ -2389,15 +2388,31 @@ private GroupedValues(
     public PCollection<KV<K, OutputT>> apply(
         PCollection<? extends KV<K, ? extends Iterable<InputT>>> input) {
 
-      final PerKeyCombineFnRunner<? super K, ? super InputT, ?, OutputT> combineFnRunner =
-          PerKeyCombineFnRunners.create(fn);
       PCollection<KV<K, OutputT>> output = input.apply(ParDo.of(
           new OldDoFn<KV<K, ? extends Iterable<InputT>>, KV<K, OutputT>>() {
             @Override
-            public void processElement(ProcessContext c) {
+            public void processElement(final ProcessContext c) {
               K key = c.element().getKey();
 
-              c.output(KV.of(key, combineFnRunner.apply(key, c.element().getValue(), c)));
+              OutputT output;
+              if (fn instanceof RequiresContextInternal) {
+                output = ((KeyedCombineFnWithContext<? super K, ? super InputT, ?, OutputT>) fn)
+                    .apply(key, c.element().getValue(), new CombineWithContext.Context() {
+                      @Override
+                      public PipelineOptions getPipelineOptions() {
+                        return c.getPipelineOptions();
+                      }
+
+                      @Override
+                      public <T> T sideInput(PCollectionView<T> view) {
+                        return c.sideInput(view);
+                      }
+                    });
+              } else {
+                output = ((KeyedCombineFn<? super K, ? super InputT, ?, OutputT>) fn)
```

Comment:
I would actually also do `else if(fn instanceof KeyedCombineFn)` here prior to cast, and then add an `else` branch with a more informative error message like "got a `CombineFn` that I don't know how to handle".